src/components/utils/ZoomImage.tsx

```diff
@@ -0,0 +1,259 @@
+'use client';
+
+import IconMinimize from '@geist-ui/icons/minimize';
+import classNames from 'classnames';
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import { tcls } from '@/lib/tailwind';
+
+import styles from './ZoomImage.module.css';
+
+/**
+ * Replacement for an <img> tag that allows zooming.
+ * The implementation uses the experimental View Transition API in Chrome for a smooth transition.
+ */
+export function ZoomImage(
+    props: React.ComponentPropsWithoutRef<'img'> & {
+        src: string;
+    },
+) {
+    const { src, alt, width } = props;
+
+    const imgRef = React.useRef<HTMLImageElement>(null);
+    const [zoomable, setZoomable] = React.useState(false);
+    const [active, setActive] = React.useState(false);
+    const [opened, setOpened] = React.useState(false);
+
+    // Only allow zooming when image will not actually be larger and on mobile
+    React.useEffect(() => {
+        if (isTouchDevice()) {
+            setZoomable(false);
+            return;
+        }
+
+        const imageWidth = typeof width === 'number' ? width : 0;
+        let viewWidth = 0;
+
+        const mediaQueryList = window.matchMedia('(min-width: 768px)');
+        const resizeObserver =
+            width && typeof ResizeObserver !== 'undefined'
+                ? new ResizeObserver((entries) => {
+                      viewWidth = entries[0]?.contentRect.width;
+                      onChange();
+                  })
+                : null;
+
+        const onChange = () => {
+            if (!mediaQueryList.matches) {
+                // Don't allow zooming on mobile
+                setZoomable(false);
+            } else if (resizeObserver && imageWidth && viewWidth && imageWidth <= viewWidth) {
+                // Image can't be zoomed if it's already rendered as it's largest size
+                setZoomable(false);
+            } else {
+                setZoomable(true);
+            }
+        };
+
+        mediaQueryList.addEventListener('change', onChange);
+        if (imgRef.current) {
+            resizeObserver?.observe(imgRef.current);
+        }
+
+        if (!resizeObserver) {
+            // When resizeObserver is available, it'll take care of calling the changelog as soon as the element is observed
+            onChange();
+        }
+
+        return () => {
+            resizeObserver?.disconnect();
+            mediaQueryList.removeEventListener('change', onChange);
+        };
+    }, [imgRef, width, setZoomable]);
```

Comment:
There is two responsibilities in this, I think it should be splitted in several useEffect, but anyway I think it's fine.

Reply:
It's fine, nothing says that useEffect should have only one responsibility, especially when controlling a single state based on multiple conditions.